bve-obj-conv/src/main.rs

```diff
@@ -0,0 +1,236 @@
+    let mut lock = stdout.lock();
+    lock.write_all(data.as_ref()).expect("Could not write to stdout");
+    lock.flush().expect("Could not flush stdout");
+}
+
+fn read_data() -> String {
+    let stdin = std::io::stdin();
+    let mut lock = stdin.lock();
+    let mut string = String::new();
+    lock.read_line(&mut string).expect("Could not read input");
+    string.pop(); // remove newline
+    if let Some(idx) = string.find('"') {
+        string = string.chars().skip(idx + 1).take_while(|c| *c != '"').collect();
+    };
+    string
+}
+
+enum State {
+    Input,
+    Wrong,
+}
+
+fn get_input<T>(prompt: impl AsRef<[u8]>, mut verification: impl FnMut(Option<PathBuf>) -> Result<T, String>) -> T {
+    let mut state = State::Input;
+    loop {
+        match state {
+            State::Input => {
+                print(prompt.as_ref());
+                let input = read_data();
+                let trimmed = input.trim();
+                let verifiable = if trimmed.is_empty() {
+                    None
+                } else {
+                    Some(PathBuf::from(trimmed))
+                };
+                match verification(verifiable) {
+                    Ok(value) => break value,
+                    Err(msg) => {
+                        state = State::Wrong;
+                        print(msg + "\n");
+                    }
+                }
+            }
+            State::Wrong => {
+                print("\nTry again:\n");
+                state = State::Input;
+            }
+        }
+    }
+}
+
+fn obj_to_csv(input_size: usize, obj: Obj<SimplePolygon>) -> String {
+    let mut output = String::from(concat!(
+        "; Generated by BVE-Reborn .obj to .csv Converter\n; bve-obj-cov v",
+        env!("CARGO_PKG_VERSION"),
+        "\n; Do not edit manually\n"
+    ));
+    output.reserve(input_size * 4);
+    for object in &obj.objects {
+        for group in &object.groups {
+            let mut translation: HashMap<(usize, Option<usize>), usize> = HashMap::new();
+            let mut vert_count = 0_usize;
+            output.push_str("\nCreateMeshBuilder\n");
+
+            if let Some(material) = &group.material {
+                if let Some(texture) = &material.map_kd {
+                    output.push_str(&format!("LoadTexture, {}\n", texture));
+                } else if let Some(color) = &material.kd {
+                    let filter_color = |color: f32| (color * 255.0).round().min(255.0) as u8;
+                    output.push_str(&format!(
+                        "SetColor, {}, {}, {}, 255\n",
+                        filter_color(color[0]),
+                        filter_color(color[1]),
+                        filter_color(color[2])
+                    ));
+                }
+            }
+
+            for polygon in &group.polys {
+                let mut indices = Vec::new();
+                for index_pair in polygon {
+                    let position_idx = index_pair.0;
```

Comment:
Can you maybe split everything in this scope block into its own function? Or otherwise break up this function

Reply:
Sure